### ISSUE TITLE
Slow autocompletion/formatting

### ISSUE THREAD

**Issue by Lucidyan** [DonJayamanne/pythonVSCode#581]
## Environment data
VS Code version:  
- Version 1.8.0
- Commit 38746938a4ab94f2f57d9e1309c51fd6fb37553d
- Date 2016-12-13T18:03:00.506Z
- Shell 1.4.6
- Renderer 53.0.2785.143
- Node 6.5.0

Python Extension version:  0.5.5
Python Version:  3.5.2
OS and version:   Ubuntu 16.10 | Linux 4.8.0-30-generic 

## Actual behavior
Very slow autocompletion. Every time I wait a few seconds before you get a list of autocompletion after ctrl + space. Also slow format is observed.

Brakes in Vim, PyCharm and other programs are not observed.

I met references of the problem here, among the problems in Git not find anything like:
https://www.reddit.com/r/Python/comments/50513t/visual_studio_code_python_extension_slow/

## Expected  behavior
Few miliseconds autocomplation/indent. Caching?

I really like your plugin and IDE, maybe I'm doing something wrong?

**Comment by DonJayamanne** [comment]
Will add support for caching in the next release.

**Comment by DonJayamanne** [comment]
Added a new setting to define the list of modules to be preloaded (fix will be available in the next release - sometime next week):
```json
"python.autoComplete.preloadModules": [
		"numpy", "pandas", "aewfaewf"
	],
```

**Comment by RamanvProjects** [comment]
This shouldn't be closed yet. I'm still seeing the same behavior.

**Comment by martinfiers** [comment]
Ah, I think it's related to the following. I was also just debugging through the preloading. I think:

    jedi.preload_module(*modulesToLoad.split(','))
Should be

    jedi.preload_module(*modulesToLoad.split(','))

(as jedi.preload_module accepts an argument list, but not a Python list). @DonJayamanne  could you have a look at this?

**Comment by RamanvProjects** [comment]
This should be re-opened. Still seeing the same behavior (5-6 seconds for tf completion and 4-5 seconds for np completion). I'll attempt to debug this later.

**Comment by bchamand** [comment]
I agree with @RamanvProjects, with tensorflow and numpy, the completion is very slow, I have to wait about 5-6 seconds.

**Comment by Lucidyan** [comment]
@DonJayamanne  
#903 looks like a duplicate. Should not they be merged?

**Comment by LordAmit** [comment]
This is interesting. Intellisense got considerably faster after first invocation when I am using the following workspace settings: 

```javascript
{
    "python.pythonPath": "${workspaceRoot}/venv/bin/python",
    "python.autoComplete.extraPaths": [
        "${workspaceRoot}/venv/lib/python2.7/site-packages"
    ],
    "python.autoComplete.preloadModules": [
        "numpy",
        "pandas",
        "matplotlib",
        "sklearn"
    ],
    "python.workspaceSymbols.exclusionPatterns": [
  ]
}
```
Of course, first invocation is really slow, takes around 7-10 seconds. 

I noticed that the default user settings exclude site-packages, which generally contain custom modules.
```
  "python.workspaceSymbols.exclusionPatterns": [
    "**/site-packages/**"
  ],
```

Is that somehow conflicting with the IntelliSense?

**Comment by outtoin** [comment]
@LordAmit it works? My vscode still slow although add ```python.autoComplete.preloadModules``` and ```python.workspaceSymbols.exclusionPatterns``` in my settings.json

**Comment by LordAmit** [comment]
@outtoin does not exactly work when you press ctrl+space / . for the first time. Takes 7-10 seconds to show the IntelliSense entries. Then, it becomes a bit quicker when showing the entries until VSCode is closed.

**Comment by LordAmit** [comment]
Hmm, this is interesting.
I am using VSCode latest version.

```
Version 1.13.1
Commit 379d2efb5539b09112c793d3d9a413017d736f89
Date 2017-06-14T18:18:03.540Z
Shell 1.6.6
Renderer 56.0.2924.87
Node 7.4.0
```
I made a few changes in my workstyle, which are"
- started using virtualenvwrapper
- several changes in user settings: 

```
{
    "sync.gist": "REMOVEDTHIS",
    "sync.lastUpload": "",
    "sync.autoDownload": false,
    "sync.autoUpload": false,
    "sync.lastDownload": "2017-06-10T11:04:28.672Z",
    "sync.forceDownload": false,
    "sync.anonymousGist": false,
    "sync.host": "",
    "sync.pathPrefix": "",
    "sync.quietSync": false,
    "workbench.iconTheme": "vs-seti",      "editor.minimap.enabled": false,
    // "python.autoComplete.extraPaths": ["/home/amit/.local/lib/python2.7/site-packages"],
    "workbench.colorTheme": "One Dark Pro",
    "todohighlight.include": "{**/*.py, **/*.js,**/*.jsx,**/*.ts,**/*.tsx,**/*.html,**/*.php,**/*.css,**/*.scss}", //A glob pattern that defines the files to search for. Only include files you need, DO NOT USE `{**/*.*}` for both permormance and avoiding binary files reason
    "todohighlight.exclude": "{**/venv/**,**/node_modules/**,**/bower_components/**,**/dist/**,**/build/**,**/.vscode/**,**/_output/**,**/*.min.*,**/*.map}",//A glob pattern that defines files and folders to exclude while listing annotations
    "todohighlight.maxFilesForSearch": 5120,//Max files for searching, mostly you don't need to configure this
    "todohighlight.toggleURI": true,
    "workbench.colorCustomizations":{
    "tab.activeBackground": "#282c34",
    "activityBar.background": "#282c34",
    "editorGroup.background": "#282c34",
    "sideBar.background": "#282c34"
    },
    "files.trimTrailingWhitespace": true,
    "editor.cursorBlinking": "phase",
    "editor.fontSize": 13.5,
    "editor.quickSuggestions": {
    "other": true,
    "comments": true,
    "strings": false
  },
  "editor.wordWrap": "on",
  "editor.renderWhitespace": "none",
  "latex-workshop.chktex.enabled": true,
  "cSpell.ignoreRegExpList": [
  "\\\\\\w*(\\[.*?\\])?(\\{.*?\\})?",
  "\\$.+?\\$"
],
"cSpell.userWords": [
  "Likert",
  "logcat"
],
"editor.multiCursorModifier": "ctrlCmd",
"workbench.editor.enablePreview": false,
"python.linting.mypyEnabled": true,
"python.formatting.formatOnSave": true,
// "python.autoComplete.extraPaths": [
//         "/home/amit/.local/lib/python2.7/site-packages/"
//     ],
"python.autoComplete.preloadModules": [
        "numpy",
        "pandas",
        "matplotlib",
        "sklearn"
],
// "python.formatting.autopep8Path": "/home/amit/.local/lib/python2.7/site-packages/autopep8"
"python.venvPath": "~/.virtualenvs"
}
```
now, I start by ```workon environmentname``` and then start by using ```code .```.

except a few modules, specifically numpy and pandas, autocompletion of third party modules installed via pip in the virtualenvs is considerably faster. 
As usual, first invocation of pandas or numpy autocmpletion is slow, takes a few seconds. Second and next autocompletion of those two becomes faster, although still slower from regular completions.

**Comment by RamanvProjects** [comment]
Desired behavior should be something like PyCharm, which initially indexes everything (takes a bit of time) and then thereafter is instant, with no startup cost upon restarting PyCharm. It also occasionally refreshes, for updated. I think this could be done by adding binary files of the appropriate jedi preloads? And instead of doing it for all modules we only do it for big ones that the user suggests (numpy, tensorflow, etc.). Thoughts?

**Comment by simtony** [comment]
Are there any solutions?
Intellisense for numpy and tensorflow is slow, even after indexing the modules.
![sep-09-2017 21-06-10](https://user-images.githubusercontent.com/17560198/30240424-dc0e4de2-95a2-11e7-82be-13e894b34486.gif)
Such a lag is really painful.

Another problem is that even with `"python.autoComplete.preloadModules":["tensorflow", "numpy"]`,
if I then use `import tensorflow as tf` to import, and type `tf.`, the modules start to reindexing.

**Comment by JasonLiuOrg** [comment]
The same as @simtony , still very slow when write the code using tensorflow and numpy.

**Comment by JCass45** [comment]
Any progress or updates with this?

**Comment by DonJayamanne** [comment]
@JCass45 Sorry to keep you waiting. I haven't had the opportunity to look into this issue just yet. However I can assure you, this is on the plate of our todo list, and will be looked at very soon.

**Comment by JCass45** [comment]
@DonJayamanne Great, thanks for the update!

**Comment by DonJayamanne** [comment]
Closing in favor of https://github.com/DonJayamanne/pythonVSCode/issues/123
They aren't related, however in order to support #123 we'll end up re-developing a language service for intellisense and the like

**Comment by fredrikaverpil** [comment]
Was `python.autoComplete.preloadModules` ever implemented or did this not make it into a release?

I notice it's not in the wiki docs under Intellisense.

**Comment by RamanvProjects** [comment]
It was implemented. Just doesn’t work very well for large modules.

**Comment by Schronuman** [comment]
I am use the latest version of the VS code and the extension. when I try to use the **numpy**, the 2-4 seconds lag still exist even though I have added it to the `python.autoComplete.preloadModules`. 
Is there any solution for such a **painful** problem?

**Comment by ozanulku** [comment]
I have the same problem. Things work fine for a lot of imports, but a few key modules run into this problem, for example: pandas.

![output](https://user-images.githubusercontent.com/8529371/42996698-0100afac-8c1d-11e8-94e1-d8018b412eae.gif)

I have tried `python.autoComplete.preloadModules` but it made no difference. I have also tried @LordAmit's suggestions but the problem persisted.

I wish this problem would be solved. Research online led me to a lot of people experiencing the same issue but not any solutions that work. I already use VS, VS Code and Android Studio regularly for different cases, I really do not want to add another ide to that list.
